File: public/samples/DataStreams/FuturesFetch.go
// FuturesFetch.go — Fetch and decode a Futures (v14) report with the Go SDK.
//
// Usage:
//   export API_KEY="..."
//   export API_SECRET="..."
//   go run FuturesFetch.go 0x000ebfe7b7560e1866ef91d4607ce5ee7474382794e5ec755f38b70b1f30e647
package main

import (
	"context"
	"fmt"
	"os"
	"time"

	streams "github.com/smartcontractkit/data-streams-sdk/go/v2"
	feed "github.com/smartcontractkit/data-streams-sdk/go/v2/feed"
	report "github.com/smartcontractkit/data-streams-sdk/go/v2/report"
	v14 "github.com/smartcontractkit/data-streams-sdk/go/v2/report/v14"
)

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintf(os.Stderr, "Usage: go run FuturesFetch.go [FeedID]\n")
		os.Exit(1)
	}
	feedIDInput := os.Args[1]

	apiKey := os.Getenv("API_KEY")
	apiSecret := os.Getenv("API_SECRET")
	if apiKey == "" || apiSecret == "" {
		fmt.Fprintf(os.Stderr, "API_KEY and API_SECRET environment variables must be set\n")
		os.Exit(1)
	}

	cfg := streams.Config{
		ApiKey:    apiKey,
		ApiSecret: apiSecret,
		RestURL:   "https://api.testnet-dataengine.chain.link",
		Logger:    streams.LogPrintf,
	}

	client, err := streams.New(cfg)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create client: %v\n", err)
		os.Exit(1)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()

	var feedID feed.ID
	if err := feedID.FromString(feedIDInput); err != nil {
		fmt.Fprintf(os.Stderr, "Invalid feed ID format '%s': %v\n", feedIDInput, err)
		os.Exit(1)
	}

	reportResponse, err := client.GetLatestReport(ctx, feedID)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to get latest report: %v\n", err)
		os.Exit(1)
	}

	// Decode the v14 (Futures) report
	decodedReport, err := report.Decode[v14.Data](reportResponse.FullReport)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to decode report: %v\n", err)
		os.Exit(1)
	}

	d := decodedReport.Data

	fmt.Printf("\nDecoded V14 Report for Stream ID %s:\n", feedIDInput)
	fmt.Println("------------------------------------------")
	fmt.Printf("Mid Price             : %s\n", d.MidPrice.String())
	fmt.Printf("Bid Price             : %s\n", d.BidPrice.String())
	fmt.Printf("Ask Price             : %s\n", d.AskPrice.String())
	fmt.Printf("Expiry Time           : %d\n", d.ExpiryTime.Unix())
	fmt.Printf("First Day of Notice   : %d\n", d.FirstDayOfNotice.Unix())
	fmt.Printf("Last Seen TimestampNs : %d\n", d.LastSeenTimestampNs.UnixNano())
	fmt.Printf("Market Status         : %d\n", d.MarketStatus)
	fmt.Printf("Contract Month        : %d\n", d.ContractMonth)
	fmt.Printf("Goldman Roll Price    : %s\n", d.GoldmanRollPrice.String())
	fmt.Printf("Current Business Day  : %d\n", d.CurrentBusinessDay)
	fmt.Printf("Interp Goldman Roll   : %s\n", d.InterpolatedGoldmanRollPrice.String())
	fmt.Println("------------------------------------------")
}

File: public/samples/DataStreams/FuturesFetch.rs
// FuturesFetch.rs — Fetch and decode a Futures (v14) report with the Rust SDK.
//
// Usage:
//   export API_KEY="..."
//   export API_SECRET="..."
//   cargo run 0x000ebfe7b7560e1866ef91d4607ce5ee7474382794e5ec755f38b70b1f30e647
use chainlink_data_streams_report::feed_id::ID;
use chainlink_data_streams_report::report::{ decode_full_report, v14::ReportDataV14 };
use chainlink_data_streams_sdk::client::Client;
use chainlink_data_streams_sdk::config::Config;
use std::env;
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: cargo run [FeedID]");
        std::process::exit(1);
    }
    let feed_id_input = &args[1];

    let api_key = env::var("API_KEY").expect("API_KEY must be set");
    let api_secret = env::var("API_SECRET").expect("API_SECRET must be set");

    let config = Config::new(
        api_key,
        api_secret,
        "https://api.testnet-dataengine.chain.link".to_string(),
        "wss://api.testnet-dataengine.chain.link/ws".to_string(),
    )
    .build()?;

    let client = Client::new(config)?;

    let feed_id = ID::from_hex_str(feed_id_input)?;

    let response = client.get_latest_report(feed_id).await?;

    let full_report = hex::decode(&response.report.full_report[2..])?;
    let (_report_context, report_blob) = decode_full_report(&full_report)?;
    let report_data = ReportDataV14::decode(&report_blob)?;

    println!("\nDecoded V14 Report for Stream ID {}:", feed_id_input);
    println!("------------------------------------------");
    println!("Mid Price             : {}", report_data.mid_price);
    println!("Bid Price             : {}", report_data.bid_price);
    println!("Ask Price             : {}", report_data.ask_price);
    println!("Expiry Time           : {}", report_data.expiry_time);
    println!("First Day of Notice   : {}", report_data.first_day_of_notice);
    println!("Last Seen TimestampNs : {}", report_data.last_seen_timestamp_ns);
    println!("Market Status         : {}", report_data.market_status);
    println!("Contract Month        : {}", report_data.contract_month);
    println!("Goldman Roll Price    : {}", report_data.goldman_roll_price);
    println!("Current Business Day  : {}", report_data.current_business_day);
    println!("Interp Goldman Roll   : {}", report_data.interpolated_goldman_roll_price);
    println!("------------------------------------------");

    Ok(())
}
